File: 14-TypeScript/Grey Multi-Tenant/packages/grey-scala/src/main/scala/com/grey/sdk/GreyClient.scala
package com.grey.sdk

import com.grey.sdk.config.GreyOptions
import com.grey.sdk.domain.*
import com.grey.sdk.grpc.*
import scala.concurrent.ExecutionContext

/** Main entry point for the Grey SDK.
  *
  * Provides access to all domain clients for interacting with the Grey API.
  *
  * Example:
  * {{{
  * import com.grey.sdk.*
  * import com.grey.sdk.config.GreyOptions
  * import scala.concurrent.ExecutionContext.Implicits.global
  *
  * val client = GreyClient(GreyOptions(host = "api.grey.example.com"))
  *
  * client.auth.login("user@example.com", "password").map { auth =>
  *   println(s"Logged in: ${auth.accessToken}")
  * }
  * }}}
  *
  * @param options
  *   Configuration options for the client
  */
final class GreyClient(val options: GreyOptions)(using ec: ExecutionContext):
  private val channel: GreyChannel = GreyChannel(options)

  // Initialize gRPC services
  private val authService: AuthGrpcService = AuthGrpcService(channel)
  private val userService: UserGrpcService = UserGrpcService(channel)
  private val projectsService: ProjectsGrpcService = ProjectsGrpcService(channel)
  private val queryService: QueryGrpcService = QueryGrpcService(channel)
  private val mutationService: MutationGrpcService = MutationGrpcService(channel)

  // Initialize domain clients
  private val _auth: AuthClient = AuthClient(channel, authService)
  private val _user: UserClient = UserClient(channel, userService)
  private val _projects: ProjectsClient = ProjectsClient(channel, projectsService)
  private val _query: QueryClient = QueryClient(channel, queryService)
  private val _mutation: MutationClient = MutationClient(channel, mutationService)

  /** Client for authentication operations (login, logout, refresh). */
  def auth: AuthClient = _auth

  /** Client for user operations (getUser). */
  def user: UserClient = _user

  /** Client for project operations (listProjects, createProject). */
  def projects: ProjectsClient = _projects

  /** Client for generic query operations. */
  def query: QueryClient = _query

  /** Client for generic mutation operations. */
  def mutation: MutationClient = _mutation

  /** Whether the client has a valid access token. */
  def isAuthenticated: Boolean = channel.isAuthenticated

  /** The current access token, if any. */
  def accessToken: Option[String] = channel.accessToken

  /** Sets the access token manually.
    *
    * Useful for restoring a session from stored credentials.
    */
  def setAccessToken(token: Option[String]): Unit =
    channel.setAccessToken(token)

  /** Gracefully shuts down the client and closes connections. */
  def shutdown(): Unit =
    _user.clearCache()
    channel.shutdown()

  /** Immediately terminates all connections. */
  def shutdownNow(): Unit =
    _user.clearCache()
    channel.shutdownNow()

object GreyClient:
  /** Creates a new Grey client with the specified options. */
  def apply(options: GreyOptions)(using ExecutionContext): GreyClient =
    new GreyClient(options)

  /** Creates a new Grey client for local development. */
  def local(port: Int = 50051)(using ExecutionContext): GreyClient =
    new GreyClient(GreyOptions.local(port))

  /** Creates a new Grey client for production. */
  def production(host: String)(using ExecutionContext): GreyClient =
    new GreyClient(GreyOptions.production(host))

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-dart/lib/src/grey_client.dart
/// Main Grey SDK client facade.
library;

import 'config/grey_options.dart';
import 'domain/auth_client.dart';
import 'domain/mutation_client.dart';
import 'domain/projects_client.dart';
import 'domain/query_client.dart';
import 'domain/user_client.dart';
import 'grpc/auth_grpc_service.dart';
import 'grpc/grey_channel.dart';
import 'grpc/mutation_grpc_service.dart';
import 'grpc/projects_grpc_service.dart';
import 'grpc/query_grpc_service.dart';
import 'grpc/user_grpc_service.dart';

/// Main entry point for the Grey SDK.
///
/// Provides access to all domain clients for interacting with the Grey API.
///
/// Example:
/// ```dart
/// final client = GreyClient(
///   GreyOptions(host: 'api.grey.example.com'),
/// );
///
/// final auth = await client.auth.login(
///   email: 'user@example.com',
///   password: 'password',
/// );
///
/// final user = await client.user.getUser();
///
/// await client.shutdown();
/// ```
class GreyClient {
  /// Creates a new Grey client with the specified options.
  GreyClient(GreyOptions options) : _channel = GreyChannel(options) {
    // Initialize gRPC services
    final authService = AuthGrpcService(_channel);
    final userService = UserGrpcService(_channel);
    final projectsService = ProjectsGrpcService(_channel);
    final queryService = QueryGrpcService(_channel);
    final mutationService = MutationGrpcService(_channel);

    // Initialize domain clients
    _auth = AuthClient(_channel, authService);
    _user = UserClient(_channel, userService);
    _projects = ProjectsClient(_channel, projectsService);
    _query = QueryClient(_channel, queryService);
    _mutation = MutationClient(_channel, mutationService);
  }

  final GreyChannel _channel;

  late final AuthClient _auth;
  late final UserClient _user;
  late final ProjectsClient _projects;
  late final QueryClient _query;
  late final MutationClient _mutation;

  /// Client for authentication operations (login, logout, refresh).
  AuthClient get auth => _auth;

  /// Client for user operations (getUser).
  UserClient get user => _user;

  /// Client for project operations (listProjects, createProject).
  ProjectsClient get projects => _projects;

  /// Client for generic query operations.
  QueryClient get query => _query;

  /// Client for generic mutation operations.
  MutationClient get mutation => _mutation;

  /// Whether the client has a valid access token.
  bool get isAuthenticated => _channel.isAuthenticated;

  /// The current access token, if any.
  String? get accessToken => _channel.accessToken;

  /// Sets the access token manually.
  ///
  /// Useful for restoring a session from stored credentials.
  void setAccessToken(String? token) {
    _channel.setAccessToken(token);
  }

  /// Gracefully shuts down the client and closes connections.
  Future<void> shutdown() async {
    _user.clearCache();
    await _channel.shutdown();
  }

  /// Immediately terminates all connections.
  Future<void> terminate() async {
    _user.clearCache();
    await _channel.terminate();
  }
}
